```{r}
library(data.table)
library(readxl)
library(magrittr)
library(tidyr)
library(ggplot2)
```

# Load Treg ATAC-seq and RNA-seq data
```{r}
atac_dt <- read_excel("../../data/atac/Treg_SATB1_vs_Treg_ctr.xlsx")  # Diff expressed atac peaks
rna_dt <- fread("../../data/deg/Treg_S_vs_A.csv")[,c(2:8, 33:41)]

norm_rna_counts <- fread("../../data/novogene_corrected_normed_gene_count.csv")[,c(1, 2:7)]
rna_dt <- merge(rna_dt, norm_rna_counts, by.x="gene_id", by.y="V1", all=F)

atac_dt <- as.data.table(separate(atac_dt, `index`, into = c("gene_chr", "peak_start", "peak_end")))

rna_dt
atac_dt
```
# Load Teff ATAC-seq and RNA-seq data
```{r}
atac_dt <- read_excel("../../data/atac/Teff_SATB1_vs_ctr.xlsx")       # Diff expressed atac peaks
rna_dt <- fread("../../data/deg/Teff_S_vs_A.csv")[,c(2:8, 33:41)]

norm_rna_counts <- fread("../../data/novogene_corrected_normed_gene_count.csv")[,c(1, 14:19)]
rna_dt <- merge(rna_dt, norm_rna_counts, by.x="gene_id", by.y="V1", all=F)

atac_dt <- as.data.table(separate(atac_dt, `...1`, into = c("gene_chr", "peak_start", "peak_end")))

rna_dt
atac_dt
```

# Filter out insignificant peaks and DEGs
```{r}
atac_dt[, peak_start := as.numeric(peak_start)]
atac_dt[, peak_end := as.numeric(peak_end)]

atac_dt <- atac_dt[padj<0.05 & abs(log2FoldChange)>.5]  
rna_dt <- rna_dt[padj<0.05 & abs(log2FoldChange)>.5]    

rna_dt <- rna_dt[baseMean > 50]
rna_dt[, gene_chr:=paste0("chr", gene_chr)] 

atac_dt
rna_dt
```

# Get transcription start sites (tss) of the DEGs
```{r}
library(biomaRt)
mart <- useMart(biomart="ENSEMBL_MART_ENSEMBL", host="grch37.ensembl.org", path="/biomart/martservice", dataset="hsapiens_gene_ensembl")

tss_dt <- getBM(attributes=c("ensembl_gene_id","chromosome_name","transcription_start_site", "transcript_start", "transcript_end"), 
  filters = 'ensembl_gene_id', values = rna_dt$gene_id, mart = mart) %>%
  as.data.table()

tss_dt <- tss_dt[, .(ensembl_gene_id, tss=transcription_start_site)]
tss_dt

rna_dt <- merge(rna_dt, tss_dt, by.x="gene_id", by.y="ensembl_gene_id", all=F)
rna_dt
```

```{r}
# Get overlapping peaks for given gene
get_overlapping_peaks <- function(gene_id, chr, tss, window_size=10000) {
  # finds overlaps 
  overlaps <- atac_dt[gene_chr == chr & tss-window_size <= peak_end & peak_start <= tss+window_size]
  
  overlaps[, overlap_gene_id := gene_id]
  
  # calculates overlap distance
  overlaps[, overlap_distance := ifelse(tss <= peak_end & peak_start <= tss, 0,    
                                        ifelse(peak_end < tss, peak_end-tss, peak_start-tss))]
  return(overlaps)
}

# Get overlapping peaks for all DEGs
overlap_calc <- function(window_size=100000) {
  # List of data tables (each DEG has a datatable with their overlaping atac peaks)
  overlaps <- list()
  
  for (i in 1:nrow(rna_dt)) {
    genes <- rna_dt
    val <- get_overlapping_peaks(genes[i, gene_id], genes[i, gene_chr], genes[i, tss], window_size=window_size)
    val$gene_log2FoldChange <- genes[i, log2FoldChange]
    overlaps <- append(overlaps, list(val))
  }
  overlaps_dt <- rbindlist(overlaps)
  return(overlaps_dt)
}

overlaps_dt <- get_overlapping_peaks(10000)
```

```{r}
ggplot(overlaps_dt, aes(log2FoldChange, gene_log2FoldChange)) +
  geom_point()

ggplot(overlaps_dt, aes(overlap_distance)) +
  geom_histogram()

cor(overlaps_dt$log2FoldChange, overlaps_dt$gene_log2FoldChange, method = "spearman")


```

```{r}
overlaps_dt <- merge(overlaps_dt, unique(rna_dt[, c(1:23)]), by.x="overlap_gene_id", by.y="gene_id", all=F)
overlaps_dt
```

```{r}
filtered_dt <- overlaps_dt[, c(1,3:10,12:18,21:27,36:41)]

names(filtered_dt) <- c("overlap_gene_id", "peak_start", "peak_end", 
                        "atac_baseMean", "atac_log2FoldChange", "atac_lfcSE", "atac_stat", "atac_pvalue", "atac_padj",
                        "Treg_AAVS1_400", "Treg_AAVS1_324", "Treg_AAVS1_348", "Treg_SATB1_400", "Treg_SATB1_324", "Treg_SATB1_348",
                        "overlap_distance",
                        "gene_baseMean", "gene_log2FoldChange", "gene_lfcSE", "gene_stat", "gene_pvalue", "gene_padj", "gene_name",
                        "Treg_A719_1", "Treg_S719_2", "Treg_A720_9", "Treg_S720_10", "Treg_A721_17", "Treg_S721_18")
filtered_dt
write.table(filtered_dt, file = "./data/treg_specg_allp.csv",
            sep = "\t", row.names = F)
```

```{r}
filtered_dt <- overlaps_dt[, c(1,3:10,12:18,21:27,36:41)]

names(filtered_dt) <- c("overlap_gene_id", "peak_start", "peak_end", 
                        "atac_baseMean", "atac_log2FoldChange", "atac_lfcSE", "atac_stat", "atac_pvalue", "atac_padj",
                        "Teff_AAVS1_400", "Teff_AAVS1_324", "Teff_AAVS1_348", "Teff_SATB1_400", "Teff_SATB1_324", "Teff_SATB1_348",
                        "overlap_distance",
                        "gene_baseMean", "gene_log2FoldChange", "gene_lfcSE", "gene_stat", "gene_pvalue", "gene_padj", "gene_name",
                        "Teff_A719_5", "Teff_S719_6", "Teff_A720_13", "Teff_S720_14", "Teff_A721_21", "Teff_S721_22")
filtered_dt
write.table(filtered_dt, file = "./data/teff_specg_allp.csv",
            sep = "\t", row.names = F)
```

